Step 70

luca-heltai opened the pull request:
Fully distributed parallel immersed boundary method, using nitsche and particle tracking.

![movie](https://user-images.githubusercontent.com/2919756/69370155-ae2ffa00-0c9d-11ea-9fc2-dee1b1d8e628.gif)

luca-heltai commented:
Only the last commit is related to this PR. All other commits are in the other Pull Requests. If you want to take a look at the structure of the tutorial, just look at the last commit (step-70 with its CMakeLists.txt), and ignore all the others. :)

bangerth commented:
@luca-heltai -- want to rebase this onto current master?

kronbichler commented:
@luca-heltai for completeness: Do you have plans to include this in the 9.2 release? If yes, please set the target milestone.

luca-heltai commented:
@blaisb, I’d love for this to be in 9.2. Are you willing to give me a hand with this?

blaisb commented:
> @blaisb, I’d love for this to be in 9.2. Are you willing to give me a hand with this?

Absolutely. I was thinking about it yesterday. We can have a short discussion  soon to break down the work? I have some time being stuck at home anyway...

masterleinad commented:
Can you please fix the indentation already and possibly rebase?

kronbichler commented on examples/step-70/doc/intro.dox:
I assume you are still working on this text? In my first iteration through the text I could not find the primal consistency terms, i.e. boundary integrals of the form `-(\textbf{v}, \textbf{n}\cdot \nabla \textbf{u})_{\Gamma}` and a similar term for the pressure `(\texbtf{n}\cdot \textbf{v}, p)_{\Gamma}` that you get from integration by parts from the strong form to the weak form. In the code I also could not find it, but I assume it is not finished because it also lacks the adjoint consistency term that is present in the formulation here.

luca-heltai replied:
The current code is working on codim 0 case, not codim 1, so there are no integration by parts terms that appear. I am the one who should have finished that part of the code and of the documentation... sorry, I'm awfully late. I'll try to have the introduction finished together with Bruno as quickly as possible.

In the mean time, I started documenting the parts of the code that are stable.

kronbichler replied:
Thanks for the explanation - I have `Gamma` as a codim 1 object so deep in my mind that I keep forgetting what I read two minutes ago :wink:. I will have a look at the code parts we already have later today.

blaisb replied:
That part of the introduction was not finished when you read it :(. Sorry, I hope that was not a loss of time for you. I will do another thorough re-read of the intro today and try to finalize the documentation, the current version of the intro should already make more sense. Thanks!

kronbichler commented on examples/step-70/step-70.cc:
```diff
@@ -0,0 +1,1238 @@
+    void initial_setup();
+    void setup_dofs();
+    void assemble_stokes_system();
+    void assemble_nitche_restriction();
```
This is just a random typo I wanted to point out:
```suggestion
    void assemble_nitsche_restriction();
```

luca-heltai replied:
thanks!

luca-heltai commented on examples/step-70/step-70.cc:
I stopped here today. We'll keep going tomorrow.

luca-heltai commented:
@gassmoeller , would you mind taking a look at this tutorial program? This is the first tutorial in which we show a possible use of ParticleHandler. 
Would you like to copy/paste a small introduction on particles in general somewhere in the tutorial? If this could convince you more, we'd be very happy to add you as a coauthor of the tutorial itself, since you are the author of ParticleHandler... :)

bangerth commented:
On 5/2/20 9:15 AM, Luca Heltai wrote:
> @gassmoeller <https://github.com/gassmoeller> , would you mind taking a look 
> at this tutorial program? This is the first tutorial in which we show a 
> possible use of ParticleHandler.
> Would you like to copy/paste a small introduction on particles in general 
> somewhere in the tutorial? If this could convince you more, we'd be very happy 
> to add you as a coauthor of the tutorial itself, since you are the author of 
> ParticleHandler... :)

I have had the thought of writing a small example for just the 
ParticleHandler. The application I had in mind is a cathode ray example where 
you have electric field that accelerates particles -- basically just a Laplace 
solver plus particles.

That isn't going to be ready for this release, but I think it would be nice to 
have. I'm just bringing this up because I'm thinking that eventually there 
will be a simpler "first" particle step (maybe step-68?).

peterrum commented on examples/step-70/step-70.cc:
The communication only happens when visualization is requested or every time step?

blaisb replied:
The communication happens every time step when the trilinos vector is used to carry out the interpolation. It was done this way to "mimic" a FSI interaction problem in which the solid parallel distributed triangulation would not be re-distributed every time step.

peterrum commented on examples/step-70/step-70.cc:
```suggestion
    // label "initial setup" Numerous other calls to this timer are made in
```

peterrum commented on examples/step-70/step-70.cc:
This structure is somewhat complicated. What you are doing. Get a particle, get its cell, get all particles in cell, and proceed with the next particle that has not been visited. I guess I know why you are doing it this way since you don't want to iterate over all cells, but only over all cells that contain particles. 

In my opinion, the code would be more simple to read if you would loop over all cells and within that cell over all particles. If this is too expensive, I would suggest to convert the shown code to some kind of iterator which only loops over cells with particles and move it into the library.

blaisb replied:
I agree that the current code is a bit complicated, but I think adding more documentation would make the loop easier to understand. In the present case, I don't think looping over all cells is the most efficient way to proceed (although it is the easiest to understand) since the majority of the fluid cells do not contain particles...

peterrum commented on examples/step-70/doc/results.dox:
```diff
@@ -0,0 +1,305 @@
+rough approximation of the hyper-singular operator for the Poisson problem
+associated to $\Gamma$, which is precisely a Dirichlet to Neumann map.
+
+<h4> Parallel Code </h4>
```
The code is already parallel?

luca-heltai replied:
yes. The code is massively paralell, in that both the solid and the fluid are p::d::t.

blaisb commented:
> On 5/2/20 9:15 AM, Luca Heltai wrote: @gassmoeller <https://github.com/gassmoeller> , would you mind taking a look at this tutorial program? This is the first tutorial in which we show a possible use of ParticleHandler. Would you like to copy/paste a small introduction on particles in general somewhere in the tutorial? If this could convince you more, we'd be very happy to add you as a coauthor of the tutorial itself, since you are the author of ParticleHandler... :)
> I have had the thought of writing a small example for just the ParticleHandler. The application I had in mind is a cathode ray example where you have electric field that accelerates particles -- basically just a Laplace solver plus particles. That isn't going to be ready for this release, but I think it would be nice to have. I'm just bringing this up because I'm thinking that eventually there will be a simpler "first" particle step (maybe step-68?).

@bangerth I am currently working on a step with one of my student (@tonielgeitani) in which an analytical solution for a vortex (Rayleigh-Kotte or Taylor-Green) is used to move mass-less particles. The vortex velocity is either prescribed using a function, or calculated at the DOFs of the background mesh and interpolated at the position of the particles. Additionally, we wanted to take  a look at load balancing since in this case there is no calculation load on the background mesh, consequently we wanted to demonstrate how to balance the parallel triangulation when only the particles are creating work. This is very different from what you have in mind, so I think both would be very compatible. However we could maybe try to uniformise the presentation?  My student is still learning so it will most likely still take a month until everything is finished, but the results are already there ( see movie ). We can move this discussion somehere else :).

![vortex](https://user-images.githubusercontent.com/7850307/80919059-f3b91e80-8d35-11ea-8133-ab311a92453e.gif)

luca-heltai commented:
>Unfortunately, the formulas have been not rendered correctly locally. I thought they are not finished yet, so I ignored them on my printed version, but notices during adding the comments in GitHub that they actually there... They would have helped... Sigh...

I'm using doxygen with mathjax, and they show up correclty. What is doxygen telling you?

> I have a general question. What you are doing during initialization is 1) to seed particles at quadrature points of the solid domain 

we simply attach a particle to each quadrature point, by looping over each cell, and collecting qpoints and weights.

> and 2) to position these particles in the right subdomain of the fluid domain.

this is done by the `insert_global_particles` which we added recently to deal.II for precisely this purpouse.

> Afterwards, the solid domain is not needed any more and the particles are moved analytically by the DisplacementFunction? Am I summarizing this correctly?

We don't do two way coupling in this example, but we don't want to make life difficult for people wanting to do this later (read: us). What we have implemented is the coupling part, and then we simply show how to use this for a simple Nitsche formulation. Two way coupling would mean that the particles information is used by the solid domain (for example, to compute a deformation, a force, etc.). The MPI vector representing the deformation can be readily used as a `MappingQEulerian` (or `MappingFEField`) object to deform the solid domain, and compute some elasticity problem on the solid domain. The code is ready for this, even if we have not explained this throughout. We'll make this clear in the introduction.

> Throughout reading the document (which caused me a lot of confusion), I thought that the deformation happens on the solid domain (by the analytical solution). 

Yes, this is correct.

> The solution is interpolated onto the particles (just as it is done the tracing case) 

this is not exact. There are two different types of interpolation here: from an analyitical function (which determines the position of the solid), and from a fluid finite element field (to determine the *velocity* of tracers). 

> and then the information is transferred to the fluid particles. This is not the case, is it? How would you implement this. 

It is already implemented, and it is in @blaisb PR #9891 (waiting to be merged after #9076). We use this for the velocity field of the tracers.

> This information would be a useful information and an important step towards FSI, wouldn't it?

It is indeed, and we show how to treat both directions (from solid to fluid and from fluid to solid) in the solid domain (solid to fluid) and in the tracers (fluid to solid).

From the way we implemented things, the interpolation from the analytical function could be readily replaced by the output of a solid computation. We did not want to complicate further the step, so the solid impeller simply rotates with a given velocity field. Adding elasticity is not the difficult part here, and we thought it would have only added noise.

I'll coordinate with @blaisb to extend the discussion in the introduction and in other various places to incorporate your comments. Thanks!

luca-heltai commented:
Please ignore the `resulsts.dox`. It still refers to step-60. Sorry. We are working on this guy here for the 3d case:

https://d2t1xqejof9utc.cloudfront.net/screenshots/pics/d3c8e271f3c81d2b011042ee74a3b1cf/large.gif

gassmoeller commented:
> @gassmoeller , would you mind taking a look at this tutorial program? This is the first tutorial in which we show a possible use of ParticleHandler.
Would you like to copy/paste a small introduction on particles in general somewhere in the tutorial? If this could convince you more, we'd be very happy to add you as a coauthor of the tutorial itself, since you are the author of ParticleHandler... :)

Hi Luca, I'd be happy to provide a little introduction to particles somewhere and I am also interested to read through the code to see how you used the particles (since its different from my usual application). Just to get a handle on this, what is your timeline for this step? Are you aiming to merge it before the hackathon?

bangerth commented:
Is this propeller going to actually be implemented in the tutorial? That would be so awesome! (We've had really good cases in recent tutorial programs! :-) )

bangerth commented:
For a more general tutorial program about particles, let's continue the discussion in #10043.

luca-heltai commented:
This was supposed to go into 9.2, but since tomorrow is the deadline for the release, I guess we'll just postpone it for later. Before the ackathon is actually a goot target time.

kronbichler commented:
I've had a short look over the tutorial program and it does not look too bad. @luca-heltai we have made last-minute additions of tutorial programs in the past, so I would not be opposed if you say polish the intro + results sections and code a bit until Tuesday and we merge then. Of course, it is completely up to you if you prefer to have a reasonably good version in the release now or postpone it and provide a more polished one in two weeks.

blaisb commented:
> I've had a short look over the tutorial program and it does not look too bad. @luca-heltai we have made last-minute additions of tutorial programs in the past, so I would not be opposed if you say polish the intro + results sections and code a bit until Tuesday and we merge then. Of course, it is completely up to you if you prefer to have a reasonably good version in the release now or postpone it and provide a more polished one in two weeks.

I do not know what  @luca-heltai  thinks, but I think  we could have a finalized version of the 2D version with an updated intro.dox and a result.dox by next Tuesday. Then the 3D finalized version would be an additional PR. What's your take on this Luca?

luca-heltai commented:
I'd be very happy to have this in 9.2. I was currently trying to get this guy to work in 3d: 

![image](https://user-images.githubusercontent.com/2919756/81442373-c0cabc80-9173-11ea-8860-12dc38a48efe.png)

but I'm still trying to understand what's wrong with it... :)

gassmoeller commented:
This is a very cool application. I looked over the introduction and code and made some smaller suggestions in luca-heltai/dealii#106. 

I have one major question, which may be similar to what @peterrum asked earlier:
You claim that you prepare the step for a two-way coupling between the flow and the solid-domain through the solid particles. But in the code it looks like you only keep track of the creator-owner relations for the tracer particles, not for the solid particles (comparing `setup_tracer_particles` vs. `setup_solid_particles`). Even when tracer particles move around they are still owned by the original process, which always has to query another process for the velocity at their current position. But the solid particles are simply generated at the correct (solid quadrature point) locations, but inside the fluid triangulation (and owned according to the partition of the fluid domain) and moved with analytical solid velocity. From the description I did not understand if this is intentional, and I would not know how to use the currently existing information to query information from the solid dof handler. I feel like either the two setup functions are mixed up, or I misunderstand something fundamentally.

luca-heltai commented:
Little preview:

![image](https://user-images.githubusercontent.com/2919756/81497002-f24e9f80-92bb-11ea-98b3-572c5ef61823.png)

jppelteret commented:
It's looking great @luca-heltai and @blaisb!

blaisb commented:
> This is a very cool application. I looked over the introduction and code and made some smaller suggestions in [luca-heltai#106](https://github.com/luca-heltai/dealii/pull/106).
> 
> I have one major question, which may be similar to what @peterrum asked earlier:
> You claim that you prepare the step for a two-way coupling between the flow and the solid-domain through the solid particles. But in the code it looks like you only keep track of the creator-owner relations for the tracer particles, not for the solid particles (comparing `setup_tracer_particles` vs. `setup_solid_particles`). Even when tracer particles move around they are still owned by the original process, which always has to query another process for the velocity at their current position. But the solid particles are simply generated at the correct (solid quadrature point) locations, but inside the fluid triangulation (and owned according to the partition of the fluid domain) and moved with analytical solid velocity. From the description I did not understand if this is intentional, and I would not know how to use the currently existing information to query information from the solid dof handler. I feel like either the two setup functions are mixed up, or I misunderstand something fundamentally.

@gassmoeller  You are actually right. I think it's my fault and I might have written it in a confusing way. The tracer particles are coupled exactly like you would do in an FSI problem (because the tracers are still handled by their original processor using a Trilinos Vector insteady of directly). However, this is currently not the case for the solid particles. We attach the information to the solid particles and we never have to bring back information from them to their original processor (or to their original triangulation). Consequently, we are displaying both of the required features for FSI, but using particle_handler. I think you understanding was correct.

tjhei commented on doc/doxygen/tutorial/tutorial.h.in:
```diff
@@ -969,7 +984,7 @@
  *       step-50,
  *       step-55,
- *       step-59
+ *       step-59,
```
remove comma

tjhei commented on doc/doxygen/tutorial/tutorial.h.in:
```diff
@@ -703,7 +712,9 @@
  *       step-50,
- *       step-55
+ *       step-55,
+ *       step-60,
```
why do you add step-60 here? As far as I can tell, step-60 does not use MPI.

luca-heltai replied:
You are right. I was thinking of the version I have on my PC.

tjhei commented on examples/step-70/CMakeLists.txt:
```diff
@@ -0,0 +1,62 @@
+#
+# Are all dependencies fulfilled?
+#
+IF(NOT ((DEAL_II_WITH_PETSC AND NOT DEAL_II_PETSC_WITH_COMPLEX) OR DEAL_II_WITH_TRILINOS) OR NOT DEAL_II_WITH_P4EST) # keep in one line
```
do you really need this or is this copy&paste?

luca-heltai replied:
Yes. We need this.

luca-heltai replied:
And yes. This is copy&paster.

tjhei commented on examples/step-70/doc/builds-on:
```diff
@@ -0,0 +1 @@
+step-6 step-22 step-32 step-55 step-60
```
```suggestion
step-32 step-60
```
redundant!

tjhei commented on examples/step-70/doc/results.dox:
```diff
@@ -0,0 +1,305 @@
+For example, you could use the following (perfectly valid) parameter file with
+this tutorial program:
+@code
+SHORTER VERSION HERE
```
incomplete

tjhei commented on examples/step-70/doc/results.dox:
```diff
@@ -0,0 +1,305 @@
+The output of the program will look like the following:
+
+@code
+
```
incomplete

tjhei commented on examples/step-70/doc/results.dox:
```diff
@@ -0,0 +1,305 @@
+create a very fast and optimized solver on a uniformly refined square or cubic
+grid, and embed the domain where you want to perform your computation using the
+technique presented here. This would require you to only have a surface
+representatio of your domain (a much cheaper and easier mesh to produce).
```
```suggestion
representation of your domain (a much cheaper and easier mesh to produce).
```

tjhei commented on examples/step-70/doc/results.dox:
```diff
@@ -0,0 +1,305 @@
+  to use a reference domain that is topologically equivalent to a the boundary
+  of a sphere.
+
+- use a displacement instead of the deformation to map $\Gamma_0$ into $\Gamma$
```
list formatting?

tjhei commented on examples/step-70/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+<i>This program was contributed by Luca Heltai (SISSA, Trieste) and Bruno Blais (Polytechnique Montréal)
+</i>
+
+<b>Change this!!!!</b>
```
:-)

kronbichler replied:
This should be fixed before the release, but it can wait for the release candidate.

luca-heltai commented:
~~~
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |  5.54e+03s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble Nitsche terms          |       101 |       111s |         2% |
| Assemble Stokes terms           |       101 |       208s |       3.8% |
| Initial setup                   |         1 |   0.00187s |         0% |
| Output fluid                    |       102 |      15.5s |      0.28% |
| Output solid particles          |       102 |      2.63s |         0% |
| Output tracer particles         |       102 |      2.49s |         0% |
| Refine                          |        20 |      18.4s |      0.33% |
| Set solid particle position     |       100 |       6.1s |      0.11% |
| Set tracer particle motion      |       101 |      10.8s |       0.2% |
| Setup dofs                      |        21 |      13.9s |      0.25% |
| Solve                           |       101 |  5.16e+03s |        93% |
+---------------------------------+-----------+------------+------------+
~~~
![movie](https://user-images.githubusercontent.com/2919756/81515856-38484980-9336-11ea-8ae2-489d0aef34c1.gif)

We need to work a bit on the solver. :)

blaisb commented:
Dear all, as I am finalizing the results for the 2D case, I would like your opinion on the animation of the results. I think I came up with a visible color scheme to illustrate the motion of both the impeller and the tracer particles (the real gif will be higher res):

![2d_tracing](https://user-images.githubusercontent.com/7850307/81615731-15f31200-93b0-11ea-9949-a8b3536863e0.gif)

Finally, the tutorial displays the initial and the final position of the setup.
Initial:
![anim 0000](https://user-images.githubusercontent.com/7850307/81615770-24412e00-93b0-11ea-99ab-80d034723c10.png)

Final:
![anim 0119](https://user-images.githubusercontent.com/7850307/81615778-286d4b80-93b0-11ea-884e-311d77111332.png)

Clearly, the time reversible character is not fully recovered, but the results are pretty good for a simulation that takes 240 seconds to run. The error is due to a combination of mesh sensitivity and time step.  @luca-heltai @gassmoeller would you have any suggestions on the color scheme or additional changes to the 2D results?

blaisb commented:
Finally i think a good comparison can be achieved by superposing the results:
![tracer_comparison](https://user-images.githubusercontent.com/7850307/81616409-69b22b00-93b1-11ea-9fa9-779912677235.png)

bangerth commented:
Have you thought about showing the velocity as either a vector field, or as streamlines, instead of just the magnitude? This would give an indication of direction of flow, which often makes it much easier on the eye.

blaisb commented:
> Have you thought about showing the velocity as either a vector field, or as streamlines, instead of just the magnitude? This would give an indication of direction of flow, which often makes it much easier on the eye.

That is a good suggestion. I tried many things that relied on a use of Glyphs or of Glyphs + streamlines However, there is a lot happening and it ends up being very stressful for the eyes (especially with so many things to track... (tracers+solid+fluid). The best i could come up using streamlines was this :
![stream_lines](https://user-images.githubusercontent.com/7850307/81625567-53af6500-93c7-11ea-8d2a-0dd8947d8e7f.png)

I think that in the original animation, the animation itself kind-off conveys the motion, but I remain open to all suggestions if you have a better idea :)

bangerth commented:
Yes, information overload is a real issue :-) If you attach a VTK file, I'll try playing with it a bit! (Not because I don't think you're not producing great graphics already, but because I enjoy these things :-) )

blaisb commented:
> Yes, information overload is a real issue :-) If you attach a VTK file, I'll try playing with it a bit! (Not because I don't think you're not producing great graphics already, but because I enjoy these things :-) )

@bangerth absolutely. I know my Paraview talents still need improvements anyway :)!
You can find the raw data on the following gist (yeah I added a tarfile there...) : https://gist.github.com/blaisb/a211c601067bb853a23c5beae3af0d9c
The file is aptly named solutions.tar.gz and it contains all the necessary vtu, pvtu and pvd files.

tamiko commented:
@luca-heltai @blaisb @gassmoeller 
We're very close to tagging rc2 this evening and we will have to make a call on this pull request soon whether it will make it into the release, or not. I know that such a decision is a tough one (in particular because you guys did a lot of last minute work on it). Because of that I wanted to ping your quickly and ask about your opinion on that matter. For reference, I think we should merge all features by Sunday (and this might be quite a tough timeline) :neutral_face: 

(The one thing I want to absolutely avoid is that someone now pulls three all-nighters in a row just for us not merging it in the end...)

luca-heltai commented:
I think the tutorial is in very good shape. In principle tt could be merged. BUT (and this is a big BUT):

This tutorial won't run smoothly with mpirun -np XX for arbitrary XX. There are at least two bugs (in the library) that prevent it from running smoothly. 

The first is #10175. The second one is https://github.com/geodynamics/aspect/issues/3354, which happens here as well. 

I can run the 2D problem with 4 processes and 3D problems with 8 processes. If I make the 2d problem run with 8 processes and the 3d problem with more than 8 processes (i.e., if there are areas where there are no particles or no solid quadrature points) one of the two bugs shows up, but not always. I've been trying to figure the first one for two nights now.

I'll need @gassmoeller help for the second one, since he's more expert than me on this one.

I won't be doing any other changes to step-70 until I figure out the two bugs. I am not opposed to merging it right after 9.2 has been released, and then trying to fix the bug, and may be make the tutorial more complete. @blaisb ? @gassmoeller ?

gassmoeller commented:
Sorry for being absent for a day, there were some other things going on. I will spend some time today and tomorrow towards checking the bugs because they may be relevant to the release as well (not sure how critical at the moment). Will report back what I find.

marcfehling commented on examples/step-70/step-70.cc:
```suggestion
    fluid_tria.signals.pre_distributed_refinement.connect([&tracer_particle_handler]() {
      tracer_particle_handler.register_store_callback_function();
    });

    fluid_tria.signals.post_distributed_refinement.connect([&tracer_particle_handler]() {
      tracer_particle_handler.register_load_callback_function(false);
    });
```

luca-heltai commented:
Currently, both bugs are fixed the two PRs by @marcfehling (#10215) and by Myself (#10174). I'm happily running step-70 now. My fix is really a hack. I have not fixed it, I simply moved around it. 

If you find a way to really fix #10174, we'd be in good shape. I'm waiting to see what #10215 says.

luca-heltai commented:
/rebuild

bangerth commented:
How about a movie like the following picture?
![visit0000](https://user-images.githubusercontent.com/7504421/82123591-6626f700-9757-11ea-910a-61b0afb6bb6d.png)

bangerth commented:
And this would give the following movie:
https://youtu.be/y4Gypj2jpXw

blaisb commented:
> And this would give the following movie:
> https://youtu.be/y4Gypj2jpXw

I love it :)!

blaisb commented:
> I think the tutorial is in very good shape. In principle tt could be merged. BUT (and this is a big BUT):
> 
> This tutorial won't run smoothly with mpirun -np XX for arbitrary XX. There are at least two bugs (in the library) that prevent it from running smoothly.
> 
> The first is #10175. The second one is [geodynamics/aspect#3354](https://github.com/geodynamics/aspect/issues/3354), which happens here as well.
> 
> I can run the 2D problem with 4 processes and 3D problems with 8 processes. If I make the 2d problem run with 8 processes and the 3d problem with more than 8 processes (i.e., if there are areas where there are no particles or no solid quadrature points) one of the two bugs shows up, but not always. I've been trying to figure the first one for two nights now.
> 
> I'll need @gassmoeller help for the second one, since he's more expert than me on this one.
> 
> I won't be doing any other changes to step-70 until I figure out the two bugs. I am not opposed to merging it right after 9.2 has been released, and then trying to fix the bug, and may be make the tutorial more complete. @blaisb ? @gassmoeller ?

Sorry for the very late reply. Other than the bugs that Luca identified that relate to parallelism, the tutorial is (in my opinion) in a good shape. Everything works and it already demonstrates a lot of new features (particle insertion, global insertion, particle-fluid coupling, immersed boundaries) which make it interesting. I am far too inexperienced to judge on this though :).

bangerth commented:
Feel free to embed the video into the results section. You can use the same blob of html used in step-58, if you adjust the link to the video appropriately.
